Getting started: your first 3D game tutorial (#4403)
godotengine/godot-docs: getting_started/first_3d_game/03.player_movement_code.rst

.. _doc_first_3d_game_player_movement:

Moving the player with code
===========================

It's time to code! We're going to use the input actions we created in the last
part to move the character.

Right-click the *Player* node and select *Attach Script* to add a new script to
it. In the popup, set the *Template* to *Empty* before pressing the *Create*
button.

|image0|

Let's start with the class's properties. We're going to define a movement speed,
a fall acceleration representing gravity, and a velocity we'll use to move the

> character.

Author: The series' pre-requisite is to have followed the getting started series, which covers scripting basics, including creating your first script. We can add a mention but that shouldn't be necessary for the user.

::

   extends KinematicBody

   # How fast the player moves in meters per second.
   export var speed = 14
   # The downward acceleration when in the air, in meters per second squared.
   export var fall_acceleration = 75

   var velocity = Vector3.ZERO

These are common properties for a moving body. The ``velocity`` is a 3D vector

> combining a speed with a direction. Here, we define it as a property because
> we're going to build upon its value frame after frame.

Author: ```suggestion
combining a speed with a direction. Here, we define it as a property because
we want to update and reuse its value across frames.
```

           direction.z += 1
       if Input.is_action_pressed("move_up"):
           direction.z -= 1

Here, we're going to make all calculations using the ``_physics_process()``
virtual function. Like ``_process()``, it allows you to update the node every
frame, but it's designed specifically for physics-related code like moving a
kinematic or rigid body.

.. seealso::

    To learn more about the difference between ``_process()`` and
    ``_physics_process()``, see :ref:`doc_idle_and_physics_processing`.

We start by initializing a ``direction`` variable to ``Vector3.ZERO``. Then, we
check if the player is pressing one or more of the ``move_*`` inputs and update
the vector's ``x`` and ``z`` components accordingly. These correspond to the
ground plane's axes.

These four conditions give us eight possibilities and eight possible directions.

In case the player presses, say, both W and D simultaneously, the vector will
have a length of about ``1.4``. But if they press a single key, it will have a
length of ``1``. We want the vector's length to be consistent. To do so, we can
call its ``normalize()`` method.

::

   #func _physics_process(delta):
       #...

>        if direction.length() > 0:

Reviewer: We can use the truthiness/falsiness of vectors instead of computing a length:

```suggestion
       if direction:
```

Author: Yes, that's the idea, clarity over conciseness, for educational purposes.

Reviewer: Would still be better as:

```suggestion
       if direction != Vector3.ZERO:
```

Author: Yes you're right. I'll change that.

           direction = direction.normalized()
           $Pivot.look_at(translation + direction, Vector3.UP)

Here, we only normalize the vector if the direction has a length greater than
zero, which means the player is pressing a direction key.

In this case, we also get the *Pivot* node and call its ``look_at()`` method.
This method takes a position in space to look at in global coordinates and the
up direction. In this case, we can use the ``Vector3.UP`` constant.

.. note::

    A node's local coordinates, like ``translation``, are relative to their
    parent. Global coordinates are relative to the world's main axes you can see
    in the viewport instead.

In 3D, the property that contains a node's position is ``translation``. By
adding the ``direction`` to it, we get a position to look at that's one meter
away from the *Player*.

Then, we update the velocity. We have to calculate the ground velocity and the

> fall speed separately. Be sure to go back one tab so the lines are inside the

Author: Done!

``_physics_process()`` function but outside the condition we just wrote.

::

       #if direction.length() > 0:
           #direction = direction.normalized()
           #$Pivot.look_at(translation + direction, Vector3.UP)

       # Ground velocity
       velocity.x = direction.x * speed
       velocity.z = direction.z * speed
       # Vertical velocity
       velocity.y -= fall_acceleration * delta
       # Moving the character
       velocity = move_and_slide(velocity, Vector3.UP)

For the vertical velocity, we subtract the fall acceleration multiplied by the
delta time every frame. Notice the use of the ``-=`` operator, which is a
shorthand for ``variable = variable - ...``.

This line of code will cause our character to fall in every frame. This may seem

> strange if it's already on the floor. But we have to do this for the character
> to collide with the ground every.

Author: ```suggestion
strange if it's already on the floor. But we have to do this for the character
to collide with the ground every frame.
```
